Example2/CPP_Abstraction/xt_table_info.cpp

```diff
@@ -666,42 +666,139 @@ compat_ipt_get_target(struct compat_ipt_entry *e)
-	}
-	return 0;
-}
+////////// C++ Abstraction ////////////////////////
+class XtTableInfo
+{
+    private:
+        /* Size per table */
+        unsigned int size;
+        /* Number of entries: FIXME. --RR */
+        unsigned int number;
+         /* Initial number of entries. Needed for module usage count */
+        unsigned int initial_entries;
+         /* Entry points and underflows */
+        unsigned int hook_entry[NF_INET_NUMHOOKS];
+        unsigned int underflow[NF_INET_NUMHOOKS];
+         /*
+             * Number of user chains. Since tables cannot have loops, at most
+             * @stacksize jumps (number of user chains) can possibly be made.
+        */
+         unsigned int stacksize;
+         void ***jumpstack;
+ 
+         unsigned char entries[0]; //__aligned(8);
+      
+      public:
+         XtTableInfo();//default constructor
+
+/* Getters */
+         unsigned int getSize(){
+            return size;
+         }
+
+         unsigned int getNumber(){
+            return number;
+         }
+
+         unsigned int getInitialEntries(){
+            return initial_entries;
+         }
+
+         unsigned int getStackSize(){
+            return stacksize;
+         }
+
+        void * getEntries(){
+            return entries;
+        }
+
+	  unsigned int * getHookEntry(){
+            return hook_entry;
+        }
+
+	 unsigned int * getUnderflow(){
+            return underflow;
+       }
+/* End Getters */
+
+/* Setters -- //Fix-Me - add contraint to setter where appropriate */
+	   void setSize(unsigned int mSize){
+		size = mSize;
+	   }
+         
+	   void setInitialEntry(unsigned int mInitial_entries){
+			initial_entries = mInitial_entries;			
+	   }//setInitialEntry
+
+	   void setNumber(unsigned int mNumber){
+		number = mNumber;
+	   }	 
+
+      /**	 
+       * @brief Set the Entry Offset object
+	 *Input Parameter:(offset, setHooks, e, info, base)
+	 * 	@param offset -- used to decrement the value of abstraction.size, abstraction.hook_entry, & abstraction.underflow
+	 * 	@param setHooks -- error check to decide if abstraction.hook_entry, & abstraction.underflow should decrement 
+	 * 	@param e -- pointer to rows of the table
+	 * 	@param info -- the network routing table itself
+	 * 	@param base -- pointer to starting row of the table
+       * 
+	 * Output Parameter:(info)	
+       *    @param info -- the modified table {abstraction.size, and/or (abstraction.hook_entry, & abstraction.underflow)}
+       * 
+       * Constraints{
+       *    "input.offset": "0 <= input.offset < sizeof(struct ipt_entry) <= abstraction.size",
+       *    "input.base": "base-pointer + total-sizeof(struct ipt_entry)  <= abstraction.size",
+       *    "abstraction.hook_entry & abstraction.underflow":  "[(hook_entry or underflow)+ base] <= sizeof(struct ipt_entry)"
+       * }
+       * 
+	 */
+      void setEntryOffset(int offset, int setHooks, const struct ipt_entry *e, const XtTableInfo *info, const void *base){
```

Comment:
This doesn't look right --- the constraint doesn't make sense. It is a very bad practice to require strict relations among values of different input parameters